### PR TITLE
Change to generate hipVersion.h

### PR DESCRIPTION
**Pull request by SarbojitAMD**
There is an ask to generate macros for hip major, minor and git commit date so that applications using hip can use the version information.

### CODE REVIEW

File: tests/src/deviceLib/hipVersion.cpp [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+ * Note : Compile time test hence always returns success while run.
+ * Intension is to make sure apilcation can access hip version.
+ */
+ int main()
+ {
+     std::cout<<"Hip major version : "<<HIP_VERSION_MAJOR<<std::endl;
+     std::cout<<"Hip minor version : "<<HIP_VERSION_MINOR<<std::endl;
+     std::cout<<"Hip patch version : "<<HIP_VERSION_PATCH<<std::endl;
+     passed();
+     return 0;
```
Comment (emankov) [inline tests/src/deviceLib/hipVersion.cpp]: Excessive return

File: tests/src/deviceLib/hipVersion.cpp [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+ }
```
Comment (emankov) [inline tests/src/deviceLib/hipVersion.cpp]: Add a new line

File: CMakeLists.txt [hunk 3]
```diff
@@ -343,6 +343,10 @@ endif()
 # Generate .hipVersion
 file(WRITE "${PROJECT_BINARY_DIR}/.hipVersion" ${_versionInfo})
 
+# Generate hipVersion.h
+set(_versionInfoHeader "//Auto-generated by cmake\n\n#define HIP_VERSION_MAJOR ${HIP_VERSION_MAJOR}\n#define HIP_VERSION_MINOR ${HIP_VERSION_MINOR}\n#define HIP_VERSION_PATCH ${HIP_VERSION_GITDATE}\n")
```
Comment (emankov) [inline CMakeLists.txt]: Add a header guard

Comment (emankov) [inline CMakeLists.txt]: Add a space: `// Auto`

Comment (emankov) [inline CMakeLists.txt]:
```c++
#define HIP_VERSION_PATCH ${HIP_VERSION_GITDATE}
```

contradicts to CMakeLists.txt:

```cmake
set(HIP_VERSION_PATCH ${HIP_VERSION_GITDATE}-${HIP_VERSION_GITHASH})
```

Comment (SarbojitAMD) [reply CMakeLists.txt]: Thanks

Comment (SarbojitAMD) [reply CMakeLists.txt]: hipcc passes only git workweek to application hence we have kept it to maintain current behavior,

File: tests/src/deviceLib/hipVersion.cpp [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+ * Intension is to make sure apilcation can access hip version.
+ */
+ int main()
+ {
+     std::cout<<"Hip major version : "<<HIP_VERSION_MAJOR<<std::endl;
+     std::cout<<"Hip minor version : "<<HIP_VERSION_MINOR<<std::endl;
+     std::cout<<"Hip patch version : "<<HIP_VERSION_PATCH<<std::endl;
+
+     passed();
+ }
```
Comment (emankov) [inline tests/src/deviceLib/hipVersion.cpp]: A new line at the end of a file meant.

File: CMakeLists.txt [hunk 5]
```diff
@@ -344,7 +344,14 @@ endif()
...
 # Generate hipVersion.h
-set(_versionInfoHeader "//Auto-generated by cmake\n\n#define HIP_VERSION_MAJOR ${HIP_VERSION_MAJOR}\n#define HIP_VERSION_MINOR ${HIP_VERSION_MINOR}\n#define HIP_VERSION_PATCH ${HIP_VERSION_GITDATE}\n")
+set(_versionInfoHeader
+"// Auto-generated by cmake\n
+#ifndef HIP_VERSION_H
+#define HIP_VERSION_H\n
+#define HIP_VERSION_MAJOR ${HIP_VERSION_MAJOR}
+#define HIP_VERSION_MINOR ${HIP_VERSION_MINOR}
+#define HIP_VERSION_PATCH ${HIP_VERSION_GITDATE}\n
+#endif")
```
Comment (emankov) [inline CMakeLists.txt]: And a new line at the end of the generated header file (after the closing bracket).

**Comment by scchan** [comment]
This change is causing hipcc to generate these warnings:
Name "main::HIP_VERSION_MINOR" used only once: possible typo at /opt/rocm/bin/hipcc line 122.
Name "main::HIP_VERSION_MAJOR" used only once: possible typo at /opt/rocm/bin/hipcc line 122.

**Comment by scchan** [comment]
`hipVersion.h` doesn't follow the naming convention in HIP, it should have been named as `hip_version.h` instead

**Comment by scchan** [comment]
>file(WRITE "${CMAKE_CURRENT_SOURCE_DIR}/include/hip/hipVersion.h" ${_versionInfoHeader})
It should generate the header in the build directory instead of the source directory because running cmake would create a new file not tracked by git

**Comment by gargrahul** [comment]
> 
> 
> This change is causing hipcc to generate these warnings:
> Name "main::HIP_VERSION_MINOR" used only once: possible typo at /opt/rocm/bin/hipcc line 122.
> Name "main::HIP_VERSION_MAJOR" used only once: possible typo at /opt/rocm/bin/hipcc line 122.

#1767 fixes theses though it needs some more cleanup.

**Pull request merged**